### PR TITLE
Refactor splitter&transform; Modify some data related config; Add external dataset.

### PR DESCRIPTION
**Pull request by rayrayraykk**
1. Refactor splitter; 
2. Modify some data-related config, like transforms and args; 
3. Add some external datasets from `torchvision`  and torchtext(datasets from `torchaudio` are coming soon).
4. Remove old `config.py` which is not used.
5. Refactor transforms.
6. Add a uni-test with MNIST and IMDB for the external dataset.

### CODE REVIEW

File: federatedscope/attack/example_attack_config/CRA_fedavg_convnet2_on_femnist.yaml [hunk 1]
```diff
@@ -16,6 +16,7 @@ data:
   batch_size: 10
   subsample: 0.0001
   num_workers: 0
+  transform: 'transforms.ToTensor(), transforms.Normalize(mean=[0.1307], std=[0.3081])'
```
Comment (joneswong) [inline federatedscope/attack/example_attack_config/CRA_fedavg_convnet2_on_femnist.yaml:19]: @xieyxclack @yxdyc @DavdGao  This kind of args is very tricky. Do you have any better solution?

File: federatedscope/core/auxiliaries/splitter_builder.py [hunk 2]
```diff
@@ -0,0 +1,42 @@
+import logging
+import federatedscope.register as register
+
+logger = logging.getLogger(__name__)
+
+
+def get_splitter(config):
+    client_num = config.federate.client_num
+    args = eval('{' + config.data.splitter_args + '}')
```
Comment (joneswong) [inline federatedscope/core/auxiliaries/splitter_builder.py:9]: @xieyxclack @yxdyc @DavdGao This is also very tricky. Users have to read the code about splitters so that they know the args. Any better solution?

File: federatedscope/attack/example_attack_config/CRA_fedavg_convnet2_on_femnist.yaml (hunk 1 above)
Comment (DavdGao) [reply federatedscope/attack/example_attack_config/CRA_fedavg_convnet2_on_femnist.yaml:19]: Maybe we can add a generator for transforms to support different backends, such as:
```
def get_transform(trans):
     if cfg.backend == 'torch':
        ...
    elif cfg.backend == 'tf':
        ...

```
`trans` is the selected transforms, e.g. `trans=["resize", "crop", "normalize", "flip"]`, that can be translated into the corresponding operations in torch/tensorflow.

File: federatedscope/cv/dataset/leaf_cv.py [hunk 3]
```diff
@@ -99,9 +97,9 @@ def __getitem__(self, index):
             index (int): Index
 
         Returns:
-            dict: {'train':[(image, target)], 
-                   'test':[(image, target)], 
-                   'val':[(image, target)]} 
+            dict: {'train':[(image, target)],
+                   'test':[(image, target)],
+                   'val':[(image, target)]}
```
Comment (joneswong) [inline federatedscope/cv/dataset/leaf_cv.py:102]: I cannot figure out the purpose of this method. It looks very strange to transform the whole dataset in a getitem method.

**Review by joneswong — COMMENTED** [review]
We need a discussion about this pr

File: federatedscope/attack/example_attack_config/CRA_fedavg_convnet2_on_femnist.yaml (hunk 1 above)
Comment (xieyxclack) [reply federatedscope/attack/example_attack_config/CRA_fedavg_convnet2_on_femnist.yaml:19]: Oh...IMP, this kind of args should not be allowed (like injection?). 
Maybe we can pre-define some widely-used operations (such as `ToTensor`, `Normalize`, 'Resize') and the config only contains the names of operations. But I am not sure such `transform` can only be used for torchvision datasets?

Comment (rayrayraykk) [reply federatedscope/attack/example_attack_config/CRA_fedavg_convnet2_on_femnist.yaml:19]: > Maybe we can add a generator for transforms to support different backends, such as:
> 
> ```
> def get_transform(trans):
>      if cfg.backend == 'torch':
>         ...
>     elif cfg.backend == 'tf':
>         ...
> ```
> 
> `trans` is the selected transforms, e.g. `trans=["resize", "crop", "normalize", "flip"]`, that can be translated into the corresponding operations in torch/tensorflow.

This has been done, see `get_transform` from `transform_builder`. We should a better way to handle the args of these transforms.

Comment (DavdGao) [reply federatedscope/attack/example_attack_config/CRA_fedavg_convnet2_on_femnist.yaml:19]: As for more complex transforms, e.g. `Normalize`, that require additional parameters (mean, std), we may preset  them in `get_data` for the common datasets like cifar10/100, femnist or celebA. 
Also, we should support customized mean/std in certain way, like`trans=["normalize:{'mean':0., 'std':1.}"]`

File: federatedscope/core/configs/cfg_data.py [hunk 4]
```diff
@@ -10,15 +10,18 @@ def extend_data_cfg(cfg):
 
     cfg.data.root = 'data'
     cfg.data.type = 'toy'
+    cfg.data.args = ''
+    cfg.data.splitter = ''
+    cfg.data.splitter_args = ''
```
Comment (DavdGao) [inline federatedscope/core/configs/cfg_data.py:15]: maybe we should add some annotations here to show the usage and some alternative options for `splitter` and `splitter_args`

**Review by DavdGao — COMMENTED** [review]
please see the detailed comments

File: federatedscope/core/auxiliaries/data_builder.py [hunk 5]
```diff
@@ -118,6 +118,238 @@ def _generate_data(client_num=5,
...
+                'test'] = torch.utils.data.dataset.random_split(
+                    data_list[0], lengths)
+
+        return data_dict
+
+    def load_torchaudio_data(name, splits=None, config=None):
+        import torchaudio
+
+        dataset_func = getattr(import_module('torchaudio.datasets'), name)
+        raise NotImplementedError
```
Comment (joneswong) [inline federatedscope/core/auxiliaries/data_builder.py:308]: what is the purpose

File: federatedscope/core/auxiliaries/transform_builder.py [hunk 6]
```diff
@@ -0,0 +1,44 @@
+from importlib import import_module
+import federatedscope.register as register
+
+
+def get_transform(config, package):
+    transform_funcs = {
```
Comment (joneswong) [inline federatedscope/core/auxiliaries/transform_builder.py:6]: We should not override the default values by `None`

File: tests/test_external_dataset.py [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+        cfg.federate.sample_client_rate = 0.2
+        cfg.federate.share_local_model = True
+        cfg.federate.online_aggr = True
+
+        cfg.data.root = 'test_data/'
+        cfg.data.type = 'MNIST@torchvision'
+        cfg.data.args = [{'download': True}]
+        cfg.data.splits = [0.6, 0.2, 0.2]
+        cfg.data.batch_size = 10
+        cfg.data.transform = [['ToTensor'],
```
Comment (joneswong) [inline tests/test_external_dataset.py:36]: I am curious about whether the these args (`transform` and `splitter_args`) can be specified in command line?

**Review by joneswong — COMMENTED** [review]
looks good but some minor issues need to be discussed/resolved

File: federatedscope/core/auxiliaries/data_builder.py [hunk 8]
```diff
@@ -118,6 +118,238 @@ def _generate_data(client_num=5,
     return data, config
 
 
+def load_external_data(config=None):
```
Comment (yxdyc) [inline federatedscope/core/auxiliaries/data_builder.py:121]: This function appears to do a variety of things (with several internal sub-functions). It would be nice to have a document at the beginning that explains its input and output and the logical work flow.

File: federatedscope/core/auxiliaries/data_builder.py [hunk 9]
```diff
@@ -118,6 +118,238 @@ def _generate_data(client_num=5,
...
+            else:
+                for item in data_iter:
+                    data.append(
+                        vocab.get_vecs_by_tokens(tokenizer(item[1]),
+                                                 lower_case_backup=True))
+                    targets.append(item[0])
+                targets = label_to_index(targets)
+            data = pad_sequence(data).transpose(0,
+                                                1)[:, :raw_args['max_len'], :]
+            data_list.append([(x, y) for x, y in zip(data, targets)])
```
Comment (yxdyc) [inline federatedscope/core/auxiliaries/data_builder.py:268]: For the [IterableDataset](https://pytorch.org/docs/stable/data.html#torch.utils.data.IterableDataset), we may need a more general and principled load function. I will add and test more speech data based on `IterableDataset` and we can improve this in future discussions.

File: federatedscope/core/auxiliaries/data_builder.py [hunk 10]
```diff
@@ -118,6 +118,238 @@ def _generate_data(client_num=5,
...
+        dataset_func = getattr(import_module('torchaudio.datasets'), name)
+        raise NotImplementedError
+
+    def load_torch_geometric_data(name, splits=None, config=None):
+        import torch_geometric
+
+        dataset_func = getattr(import_module('torch_geometric.datasets'), name)
+        raise NotImplementedError
+
+    load_data = {
```
Comment (yxdyc) [inline federatedscope/core/auxiliaries/data_builder.py:316]: It can be a global variable with the name `DATA_LOAD_FUNCS`, and be put in a more forward position.

File: federatedscope/core/auxiliaries/splitter_builder.py [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+    for func in register.splitter_dict.values():
+        splitter = func(config)
+        if splitter is not None:
+            return splitter
+    # Delay import
+    # generic splitter
+    if config.data.splitter == 'lda':
+        from federatedscope.core.splitters.generic import LDASplitter
+        splitter = LDASplitter(client_num, **args)
+    # graph splitter
```
Comment (yxdyc) [inline federatedscope/core/auxiliaries/splitter_builder.py:23]: Is it possible to make these graph splitters applicable to other data formats? That is, they can use the same splitting algorithms but with a bit different pre/post processes.

File: federatedscope/core/auxiliaries/transform_builder.py [hunk 12]
```diff
@@ -0,0 +1,44 @@
+from importlib import import_module
+import federatedscope.register as register
+
+
+def get_transform(config, package):
```
Comment (yxdyc) [inline federatedscope/core/auxiliaries/transform_builder.py:5]: Simple documentation can be added to tell users the input/output of the `transform`.

File: federatedscope/core/configs/cfg_data.py [hunk 13]
```diff
@@ -10,15 +10,18 @@ def extend_data_cfg(cfg):
 
     cfg.data.root = 'data'
     cfg.data.type = 'toy'
+    cfg.data.args = []  # args for external dataset, eg. [{'download': True}]
+    cfg.data.splitter = ''
+    cfg.data.splitter_args = []  # args for splitter, eg. [{'alpha': 0.5}]
+    cfg.data.transform = []  # transform for x, eg. [['ToTensor'], ['Normalize', {'mean': [0.1307], 'std': [0.3081]}]]
+    cfg.data.target_transform = []  # target_transform for y, use as above
+    cfg.data.pre_transform = []  # pre_transform for `torch_geometric` dataset, use as above
```
Comment (yxdyc) [inline federatedscope/core/configs/cfg_data.py:18]: for the three args, how about naming them as ``feat_transform``, ``label_transform``, and ``graph_pre_transform``?

File: federatedscope/core/splitters/__init__.py [hunk 14]
```diff
@@ -0,0 +1,3 @@
+from __future__ import absolute_import
+from __future__ import print_function
+from __future__ import division
```
Comment (yxdyc) [inline federatedscope/core/splitters/__init__.py:3]: No newline here.  Check whether the linter is used.

**Review by yxdyc — COMMENTED** [review]
Good job. Plz see the inline comments. After merging this PR, I will integrate more audio datasets from [wenet](https://github.com/wenet-e2e/wenet/tree/main/examples) based on the new dataset customization style.

File: federatedscope/core/auxiliaries/data_builder.py (hunk 9 above)
Comment (rayrayraykk) [reply federatedscope/core/auxiliaries/data_builder.py:268]: Thanks, I will add a `TODO` tag here.

File: federatedscope/core/auxiliaries/splitter_builder.py (hunk 11 above)
Comment (rayrayraykk) [reply federatedscope/core/auxiliaries/splitter_builder.py:23]: Sure, I will add and modify these splitters in the next PR.

File: federatedscope/core/configs/cfg_data.py (hunk 13 above)
Comment (rayrayraykk) [reply federatedscope/core/configs/cfg_data.py:18]: I can't agree. For the `torchvision` and `torch_geometric` dataset, we will use `**transform_funcs` as their args directly.

File: federatedscope/core/auxiliaries/data_builder.py [hunk 15]
```diff
@@ -119,6 +119,27 @@ def _generate_data(client_num=5,
...
+        data_local_dict: dict of split dataloader.
+                        Format:
+                            {
+                                'client_id': {
+                                    'train': DataLoader(),
+                                    'test': DataLoader(),
+                                    'val': DataLoader()
+                                }
+                            }
+        config: `CN` from `federatedscope/core/configs/config.py`
```
Comment (yxdyc) [inline federatedscope/core/auxiliaries/data_builder.py:139]: It is confusing for users why we take a `config`and return a `config`. It would be better by naming it `modifed_cfg` with more informative explanation.

Other modifications look good to me

**Review by joneswong — APPROVED** [review]
approved
